### DISCUSSION TITLE
Scoping patterns

### DISCUSSION THREAD

**Discussion by yann-soubeyrand** [cue-lang/cue#1856] [Q&A]
Hello,

I’d like to limit the scope of a pattern and I naively tried to do the following:
```
#Pulumi: self = {
	sources: dagger.#FS
	stack:   string

	{
		[string]: sources: self.sources

		about: pulumi.#Command & {
			name: "about"
		}

		preview: pulumi.#Preview & {
			stack: self.stack
		}

		update: pulumi.#Update & {
			stack: self.stack
		}
	}
}
```

I hoped that only `about`, `preview`, and `update` would get the `sources` field added, but it unfortunately tried to add `sources` to the other fields as well (`sources` and `stack`). Is it a bug or a current limitation? Is there a way to work around it?

**Comment by slewiskelly** [comment]
I'll use a modified version of what you've provided so that it is easier to replicate locally, without pulling in any external packages:

```cue
#Pulumi: self={
	sources: #FS
	stack:   string

	{
		[string]: sources: self.sources

		about: #Command & {
			name: "about"
		}

		preview: #Preview & {
			stack: self.stack
		}

		update: #Update & {
			stack: self.stack
		}
	}
}

#FS: {
	$dagger: fs: string
}

#Command: {
	name:    string
	sources: #FS
}

#Preview: {
	stack:   string
	sources: #FS
}

#Update: {
	stack:   string
	sources: #FS
}

p: #Pulumi & {
	sources: $dagger: fs: "./"
	stack: "prod"
}
```

IIUC from reading [the spec](https://cuelang.org/docs/references/spec/#declarations-and-scopes), it doesn't suggest that this is intended to be a new scope, though I understand this is the case in other languages.

When running, an error occurs. This is to be expected in the above example, and I might presume so in your original.

If I modify it to the following:

```cue
package test

#Pulumi: self={
	sources: #FS
	stack:   string

	_inner: {
		[string]: _ & {sources: self.sources}

		about: #Command & {
			name: "about"
		}

		preview: #Preview & {
			stack: self.stack
		}

		update: #Update & {
			stack: self.stack
		}
	}

	for k, v in _inner {"\(k)": v}
}

#FS: {
	$dagger: fs: string
}

#Command: {
	name:    string
	sources: #FS
}

#Preview: {
	stack:   string
	sources: #FS
}

#Update: {
	stack:   string
	sources: #FS
}

p: #Pulumi & {
	sources: $dagger: fs: "./"
	stack: "prod"
}
```

It yields the following:

```console
$ cue export .
{
    "p": {
        "sources": {
            "$dagger": {
                "fs": "./"
            }
        },
        "about": {
            "name": "about",
            "sources": {
                "$dagger": {
                    "fs": "./"
                }
            }
        },
        "preview": {
            "stack": "prod",
            "sources": {
                "$dagger": {
                    "fs": "./"
                }
            }
        },
        "stack": "prod",
        "update": {
            "stack": "prod",
            "sources": {
                "$dagger": {
                    "fs": "./"
                }
            }
        }
    }
}
```

I cannot say for sure that it is not a bug, nor if there is another way, but this seems to result in what you need.

**Comment by yann-soubeyrand** [comment]
Thanks for your answer @slewiskelly! If it’s not a bug, then it’s a feature request from my side ;-) The workaround works, but being able to introduce a new scope with curly brackets would be great.

**Comment by rogpeppe** [comment]
For the record, here's what I believe to be a simpler version of the issue:

```
x: {
	#Foo
	y: 134
}

#Foo: {
	[string]: string
	z: "hello"
}
```

This fails with:

```
x.y: conflicting values 134 and string (mismatched types int and string):
    ./tst.cue:2:2
    ./tst.cue:3:5
    ./tst.cue:7:12
```

because the pattern in the embedded `#Foo` struct matches all the fields in the struct it's embedded in too.
This is definitely by design, and I suspect that changing this behaviour would be a significant (and perhaps hard to justify) change to CUE's semantics, but I'd defer to @mpvl for that.

In general, "scope" is a bit of an ambiguous term in CUE - it usually refers to lexical scope - i.e. the fields/definitions that you can name directly. The scope you're talking about here is somewhat different - it's more dynamic.

For now, ISTM that the simplest and most "obvious" solution (from the point of view of maintenance) is just to unify the `sources` field explicitly:

```
#Pulumi: {
	sources: #FS
	stack:   string
	_sources: "sources": sources

	about: #Command & _sources & {
		name: "about"
	}

	preview: #Preview & _sources & {
		"stack": stack
	}

	update: #Update & _sources & {
		"stack": stack
	}
}
```

or even:

```
#Pulumi: {
	sources: #FS
	stack:   string
	_sources: "sources": sources
	_stack: "stack": stack

	about: #Command & _sources & {
		name: "about"
	}

	preview: #Preview & _sources & _stack
	update: #Update & _sources & _stack
}
```

HTH, and sorry for the slow reply - we've all been at Gophercon UK!

**Reply by yann-soubeyrand** [reply]
Thanks for the clarification @rogpeppe :wink:
